In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, precision_score, recall_score
from sklearn.preprocessing import StandardScaler

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_csv = pd.read_csv(Path("Resources/lending_data.csv"))

# Review the DataFrame
lending_data_csv

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_csv['loan_status']

# Separate the X variable, the features
X = lending_data_csv.drop(columns='loan_status')

In [4]:
# Review the y variable Series
display(y.head())
display(y.describe())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

count    77536.000000
mean         0.032243
std          0.176646
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: loan_status, dtype: float64

In [5]:
# Review the X variable DataFrame
display(X.head())
display(X.describe())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  
count     77536.000000      77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804  
std           1.904426          0.582086   8371.635077  
min           0.000000          0.000000      0.000000  
25%           3.000000          0.000000  14800.000000  
50%           4.000000          0.000000  18100.000000  
75%           4.000000          1.000000  21400.000000  
max          16.000000          3.000000  75200.000000

In [6]:
# Scale the X variable Dataframe Numerical Columns
numerical_columns = X.columns
scaler = StandardScaler()
X_scaled =  scaler.fit_transform(X[numerical_columns])

# Create Dataframe from X scaled Data
X_scaled_df = pd.DataFrame(X_scaled, columns=numerical_columns)

# Review X scaled Dataframe
X_scaled_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0       0.427304       0.426837         0.427404        0.668571   
1      -0.671487      -0.674918        -0.671552       -0.802166   
2      -0.384846      -0.370249        -0.372922       -0.344424   
3       0.427304       0.417844         0.415459        0.655346   
4       0.475078       0.456068         0.451295        0.694873   
...          ...            ...              ...             ...   
77531   4.440280       4.461737         4.464873        3.388970   
77532   3.771451       3.788317         3.783998        3.089553   
77533   3.723677       3.712992         3.712327        3.055563   
77534   3.102622       3.120518         3.115069        2.751245   
77535   2.768207       2.754015         2.756713        2.548451   

       num_of_accounts  derogatory_marks  total_debt  
0             0.616143          1.043996    0.427404  
1            -0.434049         -0.673973   -0.671552  
2            -0.434049         -0.673973   -0.372922  
3             0.616143          1.043996    0.415459  
4             0.616143          1.043996    0.451295  
...                ...               ...         ...  
77531         4.291814          2.761965    4.464873  
77532         3.766718          2.761965    3.783998  
77533         3.766718          2.761965    3.712327  
77534         3.241622          2.761965    3.115069  
77535         2.716526          2.761965    2.756713  

[77536 rows x 7 columns]

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X_scaled_df, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_classifier = LogisticRegression(solver="lbfgs", random_state=1)

# Fit the model using training data
logistic_classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
logistic_classifier_predictions = logistic_classifier.predict(X_test)

# Check Predictions return
display(logistic_classifier_predictions)

# Compare to y_test Data in Dataframe
pd.DataFrame({"Logistic Predictions":logistic_classifier_predictions, "y_test Scores":y_test})

array([0, 0, 0, ..., 0, 0, 0], shape=(19384,))

Logistic Predictions  y_test Scores
60914                     0              0
36843                     0              0
1966                      0              0
70137                     0              0
27237                     0              0
...                     ...            ...
45639                     0              0
11301                     0              0
51614                     0              0
4598                      0              0
2793                      0              0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Generate a confusion matrix for the model
log_matrix = confusion_matrix(y_test, logistic_classifier_predictions)

# display Logistic Regression Confusion Matrix
display(log_matrix)

# Clean up and 'translate' 
print("Logistic Regression model details:")
print(f"- Correctly classified {log_matrix[0,0]} points as 0 (Healthy Loans) but got {log_matrix[0,1]} points Incorrect")
print(f"- Correctly classified {log_matrix[1,1]} points as 1 (High-Risk Loan) but got {log_matrix[1,0]} points Incorrect")


array([[18652,   113],
       [   10,   609]])

Logistic Regression model details:
- Correctly classified 18652 points as 0 (Healthy Loans) but got 113 points Incorrect
- Correctly classified 609 points as 1 (High-Risk Loan) but got 10 points Incorrect


In [11]:
# Print the classification report for the model
target_names = ["Low-Risk Applicant (0)", "High-Risk Applicant (1)"]
print(classification_report(y_test, logistic_classifier_predictions, target_names=target_names))

                         precision    recall  f1-score   support

 Low-Risk Applicant (0)       1.00      0.99      1.00     18765
High-Risk Applicant (1)       0.84      0.98      0.91       619

               accuracy                           0.99     19384
              macro avg       0.92      0.99      0.95     19384
           weighted avg       0.99      0.99      0.99     19384



In [12]:
# Print Accuracy Score for the model
accuracy = accuracy_score(y_test, logistic_classifier_predictions) * 100
display(f"Model Accuracy: {round(accuracy, 2)}%")

# Print Precision Score for the model
precision = precision_score(y_test, logistic_classifier_predictions) * 100
display(f"Model Precision: {round(precision, 2)}%")

# Print Recall Score for the model
recall = recall_score(y_test, logistic_classifier_predictions) * 100
display(f"Model Recall: {round(recall, 2)}%")

'Model Accuracy: 99.37%'

'Model Precision: 84.35%'

'Model Recall: 98.38%'

### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model predicts both labels with a high level of accuracy. It achieved an **accuracy score of 99.37%**, indicating that the majority of predictions were correct overall. The **recall score of 98.38%** shows that the model is highly effective at identifying high-risk (`1`) applicants, minimizing the chances of missing default-prone applicants. Meanwhile, a **precision score of 84.35%** reflects that most loans predicted as high-risk were indeed high-risk, with a manageable rate of false positives for healthy (`0`) loans. Overall, the model performs reliably on both classes, with strong recall making it particularly useful in financial decision-making where identifying risk is the priority.


---